cirq-core/cirq/ops/parallel_gate.py

```diff
@@ -0,0 +1,155 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from typing import AbstractSet, Union, Any, Optional, TYPE_CHECKING, Dict
+
+import numpy as np
+
+from cirq import protocols, value
+from cirq.ops import raw_types
+from cirq.type_workarounds import NotImplementedType
+
+if TYPE_CHECKING:
+    import cirq
+
+
+@value.value_equality
+class ParallelGate(raw_types.Gate):
+    """Augments existing gates to be applied on one or more groups of qubits."""
+
+    def __init__(self, sub_gate: 'cirq.Gate', num_copies: int) -> None:
+        """Inits ParallelGate.
+
+        Args:
+            gate: the gate to apply.
+            num_copies: Number of copies of the gate to apply in parallel.
+        """
+        if sub_gate.num_qubits() != 1:
+            # TODO: If needed, support for multi-qubit sub_gates can be
+            # added by updating the circuit diagram plotting logic.
+            raise ValueError("gate must be a single qubit gate")
+        if not num_copies > 0:
+            raise ValueError("gate must be applied at least once.")
+        self._sub_gate = sub_gate
+        self._num_copies = num_copies
+
+    def num_qubits(self) -> int:
+        return self._sub_gate.num_qubits() * self._num_copies
+
+    @property
+    def sub_gate(self):
+        return self._sub_gate
+
+    @property
+    def num_copies(self):
+        return self._num_copies
+
+    def _decompose_(self, qubits):
+        if len(qubits) != self.num_qubits():
+            raise ValueError(f"len(qubits) should be {self.num_qubits()}")
```

Comment:
might want to include the given `len(qubits)` too.

Reply:
done.